java/ql/src/Performance/StringReplaceAllWithNonRegex.ql

```diff
@@ -0,0 +1,23 @@
+/**
+ * @id java/string-replace-all-with-non-regex
+ * @name Use of `String#replaceAll` with a first argument which is not a regular expression
+ * @description Using `String#replaceAll` is less performant than `String#replace` when the first
+ *              argument is not a regular expression.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity recommendation
+ * @tags performance
+ *       quality
+ *       external/cwe/cwe-1176
+ */
+
+import java
+
+from StringReplaceAllCall replaceAllCall, StringLiteral firstArg
+where
+  firstArg = replaceAllCall.getArgument(0) and
+  //only contains characters that could be a simple string
+  firstArg.getValue().regexpMatch("^[a-zA-Z0-9]+$")
+select replaceAllCall,
+  "This call to 'replaceAll' should be a call `replace` as its $@ is not a regular expression.",
```

Comment:
Does \` work in the alert message? `replaceAll` and `replace` should probably be quoted the same way.

Reply:
In VS Code, ` doesn't seem to do anything, so I've changed it to '.